### PR TITLE
feat: sensei scoring parity — WHEN: triggers, spec-security, Invalid level, advisory checks 16-18

### LINKED ISSUE #72: scoring: add WHEN: trigger pattern recognition
## Problem
The `triggerPatterns` in `internal/scoring/scoring.go` only recognizes `USE FOR:`, `triggers:`, etc. It does **not** recognize `WHEN:` — which is the **recommended** trigger format in sensei for cross-model compatibility.

## Expected Behavior
- Add `"when:"` to `triggerPatterns` in `internal/scoring/scoring.go`
- `WHEN:` should score at least as high as `USE FOR:` (sensei scores it higher)
- Update scoring tests to cover `WHEN:` triggers

## References
- spboyer/sensei `references/scoring.md` — "WHEN: preferred (scores higher)"
- `internal/scoring/scoring.go:76-81` — current trigger patterns

## Acceptance Criteria
- [ ] `WHEN:` recognized as a trigger pattern
- [ ] Existing tests pass
- [ ] New test cases for `WHEN:` triggers added

### LINKED ISSUE #73: scoring: add spec-security check (XML tags, reserved name prefixes)
## Problem
Sensei v1.2.0+ includes a `spec-security` check that validates:
1. No XML angle brackets (`< >`) in frontmatter values (injection risk)
2. Skill name does not use reserved prefixes (`claude-`, `anthropic-`)

Waza has no equivalent check.

## Expected Behavior
Add a `SpecSecurityChecker` to `internal/checks/` that:
- Flags XML angle brackets in frontmatter description and other string fields
- Flags skill names starting with `claude-` or `anthropic-` (reserved per Anthropic guide p31)
- Returns `StatusWarning` for violations

## References
- spboyer/sensei `references/scoring.md` — spec-security check definition
- Anthropic Complete Guide to Building Skills, p31

## Acceptance Criteria
- [ ] New `SpecSecurityChecker` in `internal/checks/`
- [ ] Detects `< >` in frontmatter
- [ ] Detects reserved name prefixes
- [ ] Tests covering both cases

### LINKED ISSUE #74: scoring: add Invalid score level for descriptions >1024 chars
## Problem
Sensei treats descriptions exceeding 1024 characters as `Invalid` (hard fail per agentskills.io spec). Waza's `SpecDescriptionChecker` only returns a `StatusWarning` and the `HeuristicScorer` has no `Invalid` adherence level.

## Expected Behavior
- Add `Invalid` as a score level in the adherence enum
- Descriptions >1024 chars should return `Invalid` before any other scoring
- The scoring algorithm should short-circuit: `if len(desc) > 1024 → Invalid`

## References
- spboyer/sensei `references/scoring.md` — scoring algorithm with Invalid level
- agentskills.io spec — 1024 char description constraint

## Acceptance Criteria
- [ ] `Invalid` adherence level added
- [ ] Scorer returns `Invalid` for >1024 char descriptions
- [ ] Tests for the boundary (1024, 1025 chars)

### LINKED ISSUE #75: scoring: add cross-model description density check (advisory 16)
## Problem
Sensei has advisory check 16 — cross-model description density — that validates:
1. Word count ≤60 (descriptions over 60 words dilute attention)
2. Lead sentence starts with unique action verb + domain
3. `WHEN:` with quoted phrases preferred over `USE FOR:`
4. Anti-trigger contamination warning (context-dependent)

Waza has none of these sub-checks.

## Expected Behavior
Add a `CrossModelDensityChecker` to `internal/checks/` that:
- Counts words in description, warns if >60
- Checks if first sentence leads with an action verb
- Prefers `WHEN:` format over `USE FOR:`
- Advisory only — does not change adherence level

## References
- spboyer/sensei `references/scoring.md` — advisory check 16
- kvenkatrajan gist on cross-model pattern matching

## Acceptance Criteria
- [ ] New checker in `internal/checks/`
- [ ] Word count check (>60 words = warning)
- [ ] Lead sentence heuristic
- [ ] Advisory status (not blocking)
- [ ] Tests

### LINKED ISSUE #76: scoring: add body structure quality check (advisory 17)
## Problem
Sensei v1.3.0 has advisory check 17 — body structure quality — checking whether the SKILL.md body follows effective instruction patterns:
1. Actionable instructions (code blocks, steps, commands)
2. Examples section (input → expected output scenarios)
3. Error handling (failure modes, recovery steps)
4. Troubleshooting (common problems, fixes)

Waza has no body quality checks.

## Expected Behavior
Add a `BodyStructureChecker` to `internal/checks/` that:
- Scans SKILL.md body (below frontmatter) for structure signals
- Detects: code blocks, numbered steps, "Example:", "Error:", "Troubleshooting:" sections
- Advisory only — surfaces recommendations, does not block

## References
- spboyer/sensei `references/scoring.md` — advisory check 17
- Anthropic Complete Guide to Building Skills — recommended body structure

## Acceptance Criteria
- [ ] New checker in `internal/checks/`
- [ ] Detects actionable instructions vs declarative-only
- [ ] Detects examples section
- [ ] Detects error handling / troubleshooting
- [ ] Advisory status
- [ ] Tests

### LINKED ISSUE #77: scoring: add progressive disclosure check (advisory 18)
## Problem
Sensei v1.3.0 has advisory check 18 — progressive disclosure — that flags SKILL.md bodies that should move detailed content to `references/`:
1. SKILL.md body exceeds 500 lines
2. Large code blocks (>50 lines) inline that could be in references
3. Detailed API reference or config docs inline

Waza has no progressive disclosure check.

## Expected Behavior
Add a `ProgressiveDisclosureChecker` to `internal/checks/` that:
- Counts lines in SKILL.md body
- Detects large inline code blocks (>50 lines)
- Suggests moving detailed content to `references/` directory
- Advisory only

## References
- spboyer/sensei `references/scoring.md` — advisory check 18
- agentskills.io spec — recommends under 500 lines

## Acceptance Criteria
- [ ] New checker in `internal/checks/`
- [ ] Line count check (>500 = warning)
- [ ] Large code block detection
- [ ] Advisory status
- [ ] Tests

### LINKED ISSUE #78: scoring: context-dependent anti-trigger risk assessment
## Problem
Waza's anti-trigger scoring uses blanket discouragement — any `DO NOT USE FOR:` gets the same warning. Sensei v1.3.0 uses **context-dependent risk** based on skill set size:

| Context | Risk Level |
|---------|------------|
| 1-5 skills, clear domain boundaries | Low — anti-triggers are fine |
| 5-15 skills, some overlap | Moderate — keywords start competing |
| 15+ skills, overlapping domains | **High** — keyword contamination is measurable |

## Expected Behavior
Update the anti-trigger scoring in `internal/scoring/scoring.go` to:
- Accept a skill count parameter (or detect from context)
- Scale the warning severity based on skill set size
- Low risk → informational note
- High risk → strong warning about keyword contamination
- Reference the kvenkatrajan empirical evidence

## References
- spboyer/sensei `references/scoring.md` — check 4 (anti-trigger detection)
- kvenkatrajan gist on 24 Azure skills keyword contamination
- Anthropic skills repo: 4/5 use positive-only routing

## Acceptance Criteria
- [ ] Anti-trigger risk scales with skill set size
- [ ] Low/Moderate/High risk levels
- [ ] Tests for each risk level
- [ ] `internal/scoring/scoring.go` updated

### PR DESCRIPTION
**Pull request by spboyer**
## Summary

Brings waza's scoring engine in line with [spboyer/sensei](https://github.com/spboyer/sensei) v1.3.0. Adds 7 features across scoring and checks:

### Changes

| Issue | Feature | Type |
|-------|---------|------|
| Closes #72 | `WHEN:` trigger pattern recognition | Scoring |
| Closes #73 | `spec-security` check (XML tags, reserved name prefixes) | Spec compliance |
| Closes #74 | `Invalid` score level for >1024 char descriptions | Scoring |
| Closes #75 | Cross-model description density check (advisory 16) — word count and action verb lead | Advisory |
| Closes #76 | Body structure quality check (advisory 17) — actionable instructions, examples, error handling | Advisory |
| Closes #77 | Progressive disclosure check (advisory 18) — body line count and large code block detection | Advisory |
| Closes #78 | Context-dependent anti-trigger risk assessment (severity scales by catalog size) | Scoring |

### Scope Notes

Advisory checks 16-18 implement the **core** sensei heuristics for each advisory category. Additional sub-checks (e.g., WHEN: vs USE FOR: preference analysis for advisory 16, inline reference-pattern detection for advisory 18) are deferred to follow-up work — this PR establishes the checker framework and the primary signals.

### Files Changed
- `internal/checks/advisory_checks.go` — New: CrossModelDensityChecker, BodyStructureChecker, ProgressiveDisclosureChecker
- `internal/checks/advisory_checks_test.go` — Tests for all 3 advisory checkers
- `internal/checks/score_checkers.go` — Register new checkers in pipeline
- `internal/checks/spec_checks.go` — New: SpecSecurityChecker (checks reserved name prefixes even without FrontmatterRaw)
- `internal/checks/spec_checks_test.go` — Tests for security checker (including nil FrontmatterRaw + reserved prefix)
- `internal/scoring/scoring.go` — WHEN: trigger, Invalid level, context-dependent anti-triggers (severity scales by context), section-bounded counting
- `internal/scoring/scoring_test.go` — Tests for all scoring changes
- `cmd/waza/dev/display_test.go` — Updated golden strings for new checkers
- `cmd/waza/dev/loop_test.go` — Updated golden strings for new checkers

All tests pass (`go test ./...` — 40 packages).

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: internal/scoring/scoring.go [hunk 1]
```diff
@@ -178,6 +228,31 @@ func containsAny(text string, patterns []string) bool {
...
+		if strings.EqualFold(h, header) {
+			continue
+		}
+		idx := strings.Index(strings.ToUpper(text[searchStart:]), strings.ToUpper(h))
+		if idx == -1 {
+			continue
+		}
+		absIdx := searchStart + idx
+		if absIdx < end {
+			end = absIdx
```
Comment (wbreza) [inline internal/scoring/scoring.go]: **[Medium]** sectionForHeader() calls strings.ToUpper(text) on every invocation. Since Score() calls it 3 times, the same text gets uppercased 3x. Consider pre-computing once in Score(). Minor optimization.

File: internal/checks/advisory_checks.go [hunk 2]
```diff
@@ -354,3 +354,276 @@ func (*OverSpecificityChecker) Check(sk skill.Skill) (*CheckResult, error) {
...
+			hasErrorHandling = true
+			break
+		}
+	}
+
+	var findings []string
+	hasActionable := hasCodeBlocks || hasNumberedSteps
+
+	if !hasActionable {
+		findings = append(findings, "body lacks actionable instructions (no code blocks, numbered steps, or commands)")
```
Comment (wbreza) [inline internal/checks/advisory_checks.go:506]: **[Medium]** The pattern for error is quite broad and will match code examples and variable names. Since this is advisory-level the impact is low, but consider heading-level patterns like ## error or ## troubleshooting.

**Review by wbreza — DISMISSED** [review]
## Code Review: PR #79 - feat: sensei scoring parity

### What Looks Good
- **All 8 prior Copilot review comments addressed** - thorough iteration
- **Comprehensive test coverage** - 487 lines of new tests with boundary cases
- **Clean architecture** - all checkers follow ComplianceChecker interface
- **sectionForHeader() isolates sections correctly** - solves trigger/anti-trigger overlap
- **Invalid adherence level short-circuit** - well-designed
- **Context-dependent anti-trigger risk** fires only when anti-triggers are present
- **skillBodyContent() prefers sk.Body** over raw parsing with test verification

### Suggestions (non-blocking)

1. **sectionForHeader() repeated strings.ToUpper()** - Consider pre-computing once. Minor perf.
2. **errorHandlingPatterns includes error** - Broad match. Consider heading-level patterns.

### Summary
| Priority | Count |
|----------|-------|
| Critical | 0 |
| High | 0 |
| Medium | 2 |
| Low | 2 |

**Overall Assessment:** Approve - solid feature PR with comprehensive testing.

File: internal/checks/advisory_checks.go [hunk 3]
```diff
@@ -354,3 +354,279 @@ func (*OverSpecificityChecker) Check(sk skill.Skill) (*CheckResult, error) {
...
+	"## error", "error handling", "## troubleshooting", "troubleshooting",
+	"common issues", "known limitations",
+	"warnings", "caveats", "note:", "important:",
+}
+
+func (*BodyStructureChecker) Check(sk skill.Skill) (*CheckResult, error) {
+	content := strings.ToLower(skillBodyContent(sk))
+
+	hasCodeBlocks := strings.Contains(content, "```")
+	hasNumberedSteps := regexp.MustCompile(`(?m)^\s*\d+\.\s+`).MatchString(content)
```
Comment (wbreza) [inline internal/checks/advisory_checks.go]: **[Medium]** Regex compiled on every `Check()` call

`regexp.MustCompile` is called inside `Check()`, recompiling the regex on every invocation. Other checkers in this file properly use package-level compiled regexes (lines 287-291).

**Suggested Fix:**
```suggestion
hasNumberedSteps := numberedStepPattern.MatchString(content)
```

Add at package level: `var numberedStepPattern = regexp.MustCompile(+"`"+`(?m)^\s*\d+\.\s+`+"`"+)`

File: internal/scoring/scoring.go [hunk 4]
```diff
@@ -186,7 +230,14 @@ func countPhrasesAfterPattern(text, pat string) int {
...
-	for _, stop := range []string{"DO NOT USE FOR:", "INVOKES:", "FOR SINGLE OPERATIONS:", "\n\n"} {
+	// Stop at any subsequent section header to avoid double-counting across sections.
+	for _, stop := range []string{
+		"USE FOR:", "WHEN:", "DO NOT USE FOR:", "NOT FOR:",
+		"TRIGGERS:", "INVOKES:", "FOR SINGLE OPERATIONS:", "\n\n",
+	} {
+		if strings.EqualFold(stop, pat) {
+			continue // don't stop at the pattern we're counting
+		}
 		if si := strings.Index(strings.ToUpper(after), strings.ToUpper(stop)); si >= 0 {
```
Comment (wbreza) [inline internal/scoring/scoring.go]: **[Medium]** Redundant `strings.ToUpper()` inside loop

`strings.ToUpper(after)` is called on each iteration of the stop-headers loop (7 headers). The same text gets uppercased up to 7 times per call, and `Score()` calls this function twice (~14 unnecessary allocations).

Consider pre-computing `upperAfter := strings.ToUpper(after)` once before the loop and updating it when `after` is truncated.

File: internal/checks/advisory_checks_test.go [hunk 5]
```diff
@@ -258,3 +259,273 @@ func TestOverSpecificityChecker(t *testing.T) {
...
+			passed:          false,
+			status:          StatusWarning,
+			expectBlockWarn: true,
+		},
+		{
+			name: "code block exactly 50 lines - OK",
+			rawContent: `---
+name: test
+---
+` + "```bash\n" + strings.Repeat("echo line\n", 48) + "```",
```
Comment (wbreza) [inline internal/checks/advisory_checks_test.go:474]: **[Medium]** Boundary test mislabeled and incomplete

This test is named ""code block exactly 50 lines - OK"" but uses `strings.Repeat(""echo line\n"", 48)` — that's 48 lines between fences, not 50. The true boundary (50 lines = pass since check is `> 50`, 51 lines = fail) is never tested.

Consider updating to `strings.Repeat(""echo line\n"", 50)` and adding a 51-line case that should fail.

File: internal/checks/advisory_checks.go [hunk 6]
```diff
@@ -354,3 +354,279 @@ func (*OverSpecificityChecker) Check(sk skill.Skill) (*CheckResult, error) {
...
+}
+
+// GetStatus implements StatusHolder.
+func (d *BodyStructureData) GetStatus() CheckStatus { return d.Status }
+
+var examplePatterns = []string{
+	"## example", "### example", "**example", "for example:",
+}
+
+var errorHandlingPatterns = []string{
```
Comment (wbreza) [inline internal/checks/advisory_checks.go:474]: **[Low]** Broad substring matching in `errorHandlingPatterns`

Patterns like `""error handling""` and `""warnings""` use `strings.Contains()` against the full body, which can match inside code blocks, variable names, or prose. Low-risk since this is advisory-level, but worth noting for future refinement.

File: internal/checks/advisory_checks.go [hunk 7]
```diff
@@ -354,3 +354,279 @@ func (*OverSpecificityChecker) Check(sk skill.Skill) (*CheckResult, error) {
...
+	Status           CheckStatus
+	WordCount        int
+	HasActionVerb    bool
+	ActionVerbIssues []string
+}
+
+// GetStatus implements StatusHolder.
+func (d *CrossModelDensityData) GetStatus() CheckStatus { return d.Status }
+
+var commonActionVerbs = []string{
```
Comment (wbreza) [inline internal/checks/advisory_checks.go:381]: **[Low]** ""when"" classified as action verb

""when"" is linguistically a conjunction/adverb, not an action verb. A description starting with `WHEN: processing files` gets credited as ""leading with an action verb."" This appears intentional per test coverage. Consider renaming the variable to `commonLeadWords` to better reflect the semantics.

File: internal/checks/advisory_checks.go (hunk 6 above)
Comment (spboyer) [reply internal/checks/advisory_checks.go:474]: By design — substring matching is intentionally broad for advisory checks. These are non-blocking warnings, so false positives are acceptable. Narrowing patterns would miss legitimate cases.

File: internal/checks/advisory_checks.go (hunk 7 above)
Comment (spboyer) [reply internal/checks/advisory_checks.go:381]: By design — 'when' is included because descriptions starting with 'WHEN:' are the primary trigger pattern we're checking for. While 'when' is linguistically a conjunction, it's the correct signal word in this domain.

File: internal/scoring/scoring.go [hunk 8]
```diff
@@ -126,12 +134,48 @@ func (h HeuristicScorer) Score(sk *skill.Skill) *ScoreResult {
...
 	result.AntiTriggerCount = countPhrasesAfterPattern(trimmedDesc, "DO NOT USE FOR:")
 
+	// Context-dependent anti-trigger risk (Issue #78)
+	// Severity scales with catalog size: High (≥15) → error, Moderate (6-14) → warning.
+	// Low (≤5) is silent — acceptable risk in small catalogs.
+	if h.SkillCount > 0 && !result.HasAntiTriggers {
+		var riskLevel, severity string
+		switch {
+		case h.SkillCount >= 15:
+			riskLevel = "High"
```
[the remark this answers is not in the document]
Comment (spboyer) [reply internal/scoring/scoring.go:164]: Acknowledged. The anti-trigger risk check is intentionally keyed off absence because missing anti-triggers is the risk being surfaced. This is a warning, not a gate.

File: internal/checks/spec_checks.go [hunk 9]
```diff
@@ -321,6 +321,94 @@ func (*SpecLicenseChecker) Check(sk skill.Skill) (*CheckResult, error) {
...
+	var violations []string
+
+	// Check for reserved name prefixes regardless of FrontmatterRaw presence,
+	// since Frontmatter.Name can be populated programmatically.
+	name := sk.Frontmatter.Name
+	if name != "" {
+		if strings.HasPrefix(name, "claude-") {
+			violations = append(violations, "name starts with reserved prefix 'claude-'")
+		}
+		if strings.HasPrefix(name, "anthropic-") {
```
[the remark this answers is not in the document]
Comment (spboyer) [reply internal/checks/spec_checks.go:341]: Acknowledged. Case-sensitive prefix matching is intentional — model names are always lowercase in practice. Adding ToLower would be defensive but unnecessary churn.

File: internal/checks/spec_checks.go [hunk 10]
```diff
@@ -321,6 +321,94 @@ func (*SpecLicenseChecker) Check(sk skill.Skill) (*CheckResult, error) {
...
+			violations = append(violations, "name starts with reserved prefix 'anthropic-'")
+		}
+	}
+
+	// Scan raw frontmatter values for XML angle brackets when available.
+	if sk.FrontmatterRaw != nil {
+		checkStringValue := func(key string, val any) {
```
[the remark this answers is not in the document]
Comment (spboyer) [reply internal/checks/spec_checks.go:348]: Acknowledged. FrontmatterRaw nil guard is correct — if frontmatter is populated programmatically without raw text, injection scanning is N/A. The parser always sets FrontmatterRaw when parsing from SKILL.md.

File: internal/checks/advisory_checks.go [hunk 11]
```diff
@@ -354,3 +357,279 @@ func (*OverSpecificityChecker) Check(sk skill.Skill) (*CheckResult, error) {
...
+	if strings.HasPrefix(rest, "\r\n") {
+		rest = rest[2:]
+	} else if strings.HasPrefix(rest, "\n") {
+		rest = rest[1:]
+	}
+	idx := strings.Index(rest, "\n---")
+	if idx < 0 {
+		return sk.RawContent
+	}
+	return rest[idx+4:]
```
[the remark this answers is not in the document]
Comment (spboyer) [reply internal/checks/advisory_checks.go:638]: Acknowledged. The delimiter matching handles the 99% case (YAML frontmatter with \n---). Edge cases with CRLF or no trailing newline are not worth the complexity for advisory checks.
